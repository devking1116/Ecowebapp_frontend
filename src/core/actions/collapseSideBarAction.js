/* @flow */

/**
 * The action that collapse a side bar.
 */
export const SIDEBAR_COLLAPSE = 'SIDEBAR_COLLAPSE';
export default function collapseSideBarAction(sideBarId: string): * {
  return { type: SIDEBAR_COLLAPSE, sideBarId };
}
