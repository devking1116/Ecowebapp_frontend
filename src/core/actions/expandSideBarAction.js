/* @flow */

/**
 * The action that will expand the side bar.
 */
export const SIDEBAR_EXPAND = 'SIDEBAR_EXPAND';
export default function expandSideBarAction(sideBarId: string): * {
  return { type: SIDEBAR_EXPAND, sideBarId };
}
