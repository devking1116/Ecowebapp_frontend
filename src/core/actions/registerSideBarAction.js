/* @flow */

/**
 * Create an action that will register a new side bar to the sideBarReducer.
 */
export const SIDEBAR_REGISTER = 'SIDEBAR_REGISTER';
export default function registerSideBarAction(
  sideBarId: string,
  initialExpandedWidth: number | string,
  defaultExpanded?: bool
): * {
  return {
    type: SIDEBAR_REGISTER,
    sideBarId,
    expandedWidth: initialExpandedWidth,
    defaultExpanded,
  };
}
