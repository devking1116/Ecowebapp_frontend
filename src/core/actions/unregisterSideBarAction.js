/* @flow */

/**
 * Create an action that will register a new side bar to the sideBarReducer.
 */
export const SIDEBAR_UNREGISTER = 'SIDEBAR_UNREGISTER';
export default function unregisterSideBarAction(sideBarId: string): * {
  return {
    type: SIDEBAR_UNREGISTER,
    sideBarId,
  };
}
