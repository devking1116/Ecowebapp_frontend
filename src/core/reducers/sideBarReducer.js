/* @flow */
import { SIDEBAR_COLLAPSE } from '../actions/collapseSideBarAction';
import { SIDEBAR_EXPAND } from '../actions/expandSideBarAction';
import { SIDEBAR_REGISTER } from '../actions/registerSideBarAction';
import { SIDEBAR_UNREGISTER } from '../actions/unregisterSideBarAction';

export type SideBarState = {
  expandedWidth: number | string,
  isExpanded: bool,
};

export type State = { [key: string]: SideBarState };

/**
 * A reducer to control weather or not a component is expaned or collapsed.
 */
export const SIDEBAR_REDUCER_STATE = 'SIDEBAR_REDUCER_STATE';
export default function sideBarReducer(state: State = { }, action: Object): State {
  switch(action.type) {
    case SIDEBAR_REGISTER:
      if(!action.sideBarId) {
        throw new Error('Side Bar must be given to unregister');
      }
      if(typeof action.expandedWidth !== 'number' && typeof action.expandedWidth !== 'string') {
        throw new Error('Side Bar expandedWidth must be number or string');
      }

      return {
        ...state,
        [action.sideBarId]: {
          expandedWidth: action.expandedWidth,
          isExpanded: action.defaultExpanded? true: false,
        },
      };
    case SIDEBAR_UNREGISTER:
      if(!action.sideBarId)         throw new Error('Side Bar must be given to unregister');

      return {
        ...state,
        [action.sideBarId]: undefined,
      };
    case SIDEBAR_EXPAND:
    case SIDEBAR_COLLAPSE:
      if(!action.sideBarId)         throw new Error('Side Bar must be given to update');
      if(!state[action.sideBarId])  throw new Error('Side Bar must be registered to update');

      return {
        ...state,
        [action.sideBarId]: {
          ...state[action.sideBarId],
          isExpanded: action.type === SIDEBAR_EXPAND? true: false,
        },
      };
    default:
      return state;
  }
}
