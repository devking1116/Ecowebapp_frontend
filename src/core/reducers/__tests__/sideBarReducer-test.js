import collapseSideBarAction from '../../actions/collapseSideBarAction';
import expandSideBarAction from '../../actions/expandSideBarAction';
import registerSideBarAction from '../../actions/registerSideBarAction';
import unregisterSideBarAction from '../../actions/unregisterSideBarAction';
import sideBarReducer from '../sideBarReducer';

it('has no registered side bars in its inital state', () => {
  const initalState = sideBarReducer(undefined, {});

  expect(Object.keys(initalState).length).toEqual(0);
});

it('can register a new side bar with the given name, width (in its collapsed state) and default expanded state', () => {
  const TEST_SIDEBAR = 'TEST_SIDEBAR';
  const width = 200;
  const defaultExpanded = true;
  const updatedState = sideBarReducer({}, registerSideBarAction(TEST_SIDEBAR, width, defaultExpanded))

  expect(updatedState[TEST_SIDEBAR]).toBeDefined();

  const currSideBarState = updatedState[TEST_SIDEBAR];
  expect(currSideBarState.expandedWidth).toEqual(width);
  expect(currSideBarState.isExpanded).toEqual(defaultExpanded);
});

it('can unregister a previously registered side bar', () => {
  const TEST_SIDEBAR = 'TEST_SIDEBAR';
  const updatedState = sideBarReducer({}, unregisterSideBarAction(TEST_SIDEBAR))

  expect(updatedState[TEST_SIDEBAR]).toBeUndefined();
});

it('will throw an error when trying to unregister an unregistered side bar', () => {
  const TEST_SIDEBAR = 'TEST_SIDEBAR';
  const updatedState = sideBarReducer({}, unregisterSideBarAction(TEST_SIDEBAR))

  expect(updatedState[TEST_SIDEBAR]).toBeUndefined();
});

it('can expaned a registed side bar', () => {
  const TEST_SIDEBAR = 'TEST_SIDEBAR';
  const updatedState = sideBarReducer(
    { [TEST_SIDEBAR]: { isExpanded: false } },
    expandSideBarAction(TEST_SIDEBAR),
  );

  expect(updatedState[TEST_SIDEBAR].isExpanded).toBe(true);
});

it('will throw an error when trying to expaned a unregisted side bar', () => {
  const updatedStateFn = () => sideBarReducer({ }, expandSideBarAction('TEST_SIDEBAR'));

  expect(updatedStateFn).toThrowErrorMatchingSnapshot();
});

it('can collapse a registed side bar', () => {
  const TEST_SIDEBAR = 'TEST_SIDEBAR';
  const updatedState = sideBarReducer(
    { [TEST_SIDEBAR]: { isExpanded: false } },
    collapseSideBarAction(TEST_SIDEBAR),
  );

  expect(updatedState[TEST_SIDEBAR].isExpanded).toBe(false);
});

it('will throw an error when trying to collapse a unregisted side bar', () => {
  const updatedStateFn = () => sideBarReducer({ }, collapseSideBarAction('TEST_SIDEBAR'));

  expect(updatedStateFn).toThrowErrorMatchingSnapshot();
});
