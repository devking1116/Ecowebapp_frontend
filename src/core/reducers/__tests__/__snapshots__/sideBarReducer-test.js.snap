exports[`test will throw an error when trying to collapse a unregisted side bar 1`] = `"Side Bar must be registered to update"`;

exports[`test will throw an error when trying to expaned a unregisted side bar 1`] = `"Side Bar must be registered to update"`;
